### PR TITLE
Implement Deploy and Execute CLI commands

### PR DESCRIPTION
**Pull request by raychu86**
## Motivation

This PR implements the first iteration of deploying and executing program on the network. 

Deployment and execution of programs is done via four new `developer` CLI commands in `snarkOS` - `decrypt`, `deploy`, `execute`, and `scan`. 

These CLI commands currently live in snarkOS, but can also be migrated to the Aleo SDK.

*Note: All the operations are done client-side and do not require sending private keys or view keys to third parties.*

### 1. Install snarkOS

```
cd snarkOS
git checkout feat/deploy-execute
cargo install --path .
```

### 2. Run the node in development mode

```
snarkos start --nodisplay --dev 0 --beacon ""
```

### 3. Scan the node for spendable records

This will likely be the view key associated with the development beacon.

```
snarkos developer scan -v <VIEW_KEY> --start 0 --end 1 --endpoint "http://localhost:3030"
```

### 4. Transfer credits (execute the `credits.aleo` program)

Transfer credits to another account so it can't be spent by the beacon producing blocks. 

(The beacon is constantly spending records to generate new blocks, so there is a chance your record is spent before doing this step)

```
snarkos developer execute credits.aleo transfer <INPUT_RECORD> <RECIPIENT_ADDRESS> "<AMOUNT_TO_TRANSFER>u64" --private-key <PRIVATE_KEY> --query "http://localhost:3030" --broadcast "http://localhost:3030/testnet3/transaction/broadcast" 
```

This step can be replaced with a faucet.

### 5. Deploy a program

Deploy a program with an unspent record. 

```
snarkos developer deploy fibonacci.aleo --private-key <PRIVATE_KEY> --query "http://localhost:3030" --path "../leo/examples/fibonacci/build/" --broadcast "http://localhost:3030/testnet3/transaction/broadcast" --fee 600000 --record <INPUT_RECORD>
```

### 6. Execute a function of a deployed program

```
snarkos developer execute fibonacci.aleo fibonacci "1u8" --private-key <PRIVATE_KEY> --query "http://localhost:3030" --broadcast "http://localhost:3030/testnet3/transaction/broadcast" 
```

*NOTE: Fees (in gates) must be greater than the transaction size in bytes. Fees can be excluded from execution transactions, but if one is specified, it must follow the above rule. *

### CODE REVIEW

File: cli/src/commands/developer/deploy.rs [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+    /// A path to a directory containing a manifest file. Defaults to the current working directory.
+    #[clap(long, help = "A path to a directory containing a manifest file")]
+    path: Option<String>,
+    /// The private key used to generate the deployment.
+    #[clap(short = 'p', long, help = "The private key used to generate the deployment")]
+    private_key: String,
+    /// The endpoint to query node state from.
+    #[clap(short = 'q', long, help = "The endpoint to query node state from")]
+    query: String,
+    // TODO (raychu86): Update the default.
```
Comment (ljedrz) [inline cli/src/commands/developer/deploy.rs]: has this been updated yet?

File: cli/src/commands/developer/deploy.rs [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+            std::fs::write(file_path, deployment_bytes)?;
+        }
+
+        // Determine if the transaction should be broadcast or displayed to user.
+        Developer::handle_transaction(self.broadcast, self.display, deployment, self.program_id.to_string())
+    }
+}
+
+#[cfg(test)]
+mod tests {
```
Comment (ljedrz) [inline cli/src/commands/developer/deploy.rs]: are we still planning to have these tests?

File: cli/src/commands/developer/execute.rs [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+        // Determine if the transaction should be broadcast or displayed to user.
+        Developer::handle_transaction(self.broadcast, self.display, execution, locator.to_string())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    // #[test]
```
Comment (ljedrz) [inline cli/src/commands/developer/execute.rs]: ditto

**Review by ljedrz — COMMENTED** [review]
LGTM so far :+1:

File: cli/src/commands/developer/decrypt.rs [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+
+/// Decrypts a record ciphertext.
+#[derive(Debug, Parser)]
+pub struct Decrypt {
+    /// The record ciphertext to decrypt.
+    #[clap(short = 'c', long, help = "The record ciphertext to decrypt.")]
+    pub ciphertext: String,
+    /// The view key used to decrypt the record ciphertext.
+    #[clap(short = 'v', long, help = "The view key used to decrypt the ciphertext.")]
+    pub view_key: String,
```
Comment (iamalwaysuncomfortable) [inline cli/src/commands/developer/decrypt.rs]: Clap automagically creates the short from the first letter of the variable, and the comment acts as the help, so doesn't need to be explicitly specified if you don't want.

```suggestion
    /// The record ciphertext to decrypt.
    #[clap(short, long)]
    pub ciphertext: String,
    /// The view key used to decrypt the record ciphertext.
    #[clap(short, long)]
    pub view_key: String,
```

File: cli/src/commands/developer/deploy.rs (hunk 1 above)
Comment (iamalwaysuncomfortable) [inline cli/src/commands/developer/deploy.rs]:
```suggestion
    /// The name of the program to deploy.
    #[clap(parse(try_from_str))]
    program_id: ProgramID<CurrentNetwork>,
    /// A path to a directory containing a manifest file. Defaults to the current working directory.
    #[clap(long)]
    path: Option<String>,
    /// The private key used to generate the deployment.
    #[clap(short = 'p', long, help = "The private key used to generate the deployment")]
    private_key: String,
    /// The endpoint to query node state from.
    #[clap(short, long)]
    query: String,
    // TODO (raychu86): Update the default.
```

File: cli/src/commands/developer/deploy.rs [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+    /// The endpoint to query node state from.
+    #[clap(short = 'q', long, help = "The endpoint to query node state from")]
+    query: String,
+    // TODO (raychu86): Update the default.
+    /// The deployment fee in gates.
+    #[clap(short, long, help = "The deployment fee in gates, defaults to 0")]
+    fee: Option<u64>,
+    /// The record to spend the fee from.
+    #[clap(short, long, help = "The record to spend the fee from")]
+    record: String,
```
Comment (iamalwaysuncomfortable) [inline cli/src/commands/developer/deploy.rs]:
```suggestion
    /// The deployment fee in gates, defaults to 0
    #[clap(short, long)]
    fee: Option<u64>,
    /// The record to spend the fee from.
    #[clap(short, long)]
    record: String,
```

File: cli/src/commands/developer/deploy.rs [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+    /// Display the generated transaction.
+    #[clap(
+        short,
+        long,
+        default_value = "false",
+        help = "Display the generated transaction",
+        conflicts_with = "broadcast"
+    )]
+    display: bool,
```
Comment (iamalwaysuncomfortable) [inline cli/src/commands/developer/deploy.rs]: Bool params in clap automatically default to false.

```suggestion
    /// Display the generated transaction.
    #[clap(
        short,
        long,
        conflicts_with = "broadcast"
    )]
    display: bool,
```

File: cli/src/commands/developer/deploy.rs [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+    #[clap(short, long, help = "The endpoint used to broadcast the generated transaction", conflicts_with = "display")]
+    broadcast: Option<String>,
+    #[clap(long, help = "Store generated deployment transaction to a local file")]
+    store: Option<String>,
```
Comment (iamalwaysuncomfortable) [inline cli/src/commands/developer/deploy.rs]:
```suggestion
    /// The endpoint used to broadcast the generated transaction
    #[clap(short, long, conflicts_with = "display")]
    broadcast: Option<String>,
    /// Store generated deployment transaction to a local file
    #[clap(long)]
    store: Option<String>,
```

File: cli/src/commands/developer/deploy.rs [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+    program_id: ProgramID<CurrentNetwork>,
+    /// A path to a directory containing a manifest file. Defaults to the current working directory.
+    #[clap(long, help = "A path to a directory containing a manifest file")]
+    path: Option<String>,
+    /// The private key used to generate the deployment.
+    #[clap(short = 'p', long, help = "The private key used to generate the deployment")]
+    private_key: String,
+    /// The endpoint to query node state from.
+    #[clap(short = 'q', long, help = "The endpoint to query node state from")]
+    query: String,
```
Comment (iamalwaysuncomfortable) [inline cli/src/commands/developer/deploy.rs]:
```suggestion
    /// The endpoint to query node state from.
    #[clap(short, long)]
    query: String,
```

File: cli/src/commands/developer/deploy.rs [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+    /// The name of the program to deploy.
+    #[clap(parse(try_from_str), help = "The ID of the program to deploy")]
+    program_id: ProgramID<CurrentNetwork>,
+    /// A path to a directory containing a manifest file. Defaults to the current working directory.
+    #[clap(long, help = "A path to a directory containing a manifest file")]
+    path: Option<String>,
+    /// The private key used to generate the deployment.
+    #[clap(short = 'p', long, help = "The private key used to generate the deployment")]
+    private_key: String,
+    /// The endpoint to query node state from.
```
Comment (iamalwaysuncomfortable) [inline cli/src/commands/developer/deploy.rs]:
```suggestion
    /// The name of the program to deploy.
    #[clap(parse(try_from_str))]
    program_id: ProgramID<CurrentNetwork>,
    /// A path to a directory containing a manifest file. Defaults to the current working directory.
    #[clap(long)]
    path: Option<String>,
    /// The private key used to generate the deployment.
    #[clap(short = 'p', long)]
    private_key: String,
    /// The endpoint to query node state from.
```

File: cli/src/commands/developer/execute.rs [hunk 10]
```diff
@@ -0,0 +1,173 @@
...
+        default_value = "true",
+        help = "Display the generated transaction",
+        conflicts_with = "broadcast"
+    )]
+    display: bool,
+    #[clap(short, long, help = "The endpoint used to broadcast the generated transaction", conflicts_with = "display")]
+    broadcast: Option<String>,
+    #[clap(long, help = "Store generated deployment transaction to a local file")]
+    store: Option<String>,
+}
```
Comment (iamalwaysuncomfortable) [inline cli/src/commands/developer/execute.rs]:
```suggestion
#[derive(Debug, Parser)]
pub struct Execute {
    /// The program identifier.
    #[clap(parse(try_from_str))]
    program_id: ProgramID<CurrentNetwork>,
    /// The name of the function.
    #[clap(parse(try_from_str))]
    function: Identifier<CurrentNetwork>,
    /// The function inputs.
    #[clap(parse(try_from_str))]
    inputs: Vec<Value<CurrentNetwork>>,
    /// The private key used to generate the execution.
    #[clap(short = 'p', long)]
    private_key: String,
    /// The endpoint to query node state from.
    #[clap(short, long)]
    query: String,
    /// The deployment fee in gates, defaults to 0.
    #[clap(short, long)]
    fee: Option<u64>,
    /// The record to spend the fee from.
    #[clap(short, long)]
    record: Option<String>,
    /// Display the generated transaction.
    #[clap(
        short,
        long,
        default_value = "true",
        conflicts_with = "broadcast"
    )]
    display: bool,
    /// The endpoint used to broadcast the generated transaction
    #[clap(short, long, conflicts_with = "display")]
    broadcast: Option<String>,
    /// Store generated deployment transaction to a local file
    #[clap(long)]
    store: Option<String>,
}
```

File: cli/src/commands/developer/mod.rs [hunk 11]
```diff
@@ -0,0 +1,159 @@
...
+pub enum Developer {
+    /// Decrypt a ciphertext
+    Decrypt(Decrypt),
+    /// Deploy a program.
+    Deploy(Deploy),
+    /// Execute a program function.
+    Execute(Execute),
+    /// Scan the node for records.
+    Scan(Scan),
+}
```
Comment (iamalwaysuncomfortable) [inline cli/src/commands/developer/mod.rs]: We should potentially consider a simple transfer function here. Also this functionality I think can go almost verbatim into the Aleo CLI. We should talk with @collinc97 about how to properly do this. 

The following are feasible paths:
1. All of the "Developer" functions are moved into the Aleo CLI and maintained there, while the SnarkOS node management functions remain here
2. All functions within "SnarkOS" are moved to the Aleo CLI and effectively renamed/maintained there. And exist as a multi-repo.
3. The Developer functions here are duplicated and potentially have slightly different functionality in the Aleo CLI

File: cli/src/commands/developer/mod.rs [hunk 12]
```diff
@@ -0,0 +1,159 @@
...
+            manifest.program_id() == &program_id,
+            "The program name in the manifest file does not match the specified program name"
+        );
+
+        // Load the package.
+        let package = Package::open(&directory)?;
+        // Load the main program.
+        let program = package.program();
+        // Prepare the imports directory.
+        let imports_directory = package.imports_directory();
```
Comment (iamalwaysuncomfortable) [inline cli/src/commands/developer/mod.rs:101]: Eventually we may want to add a flag to download an already deployed program in addition to searching locally.

File: cli/src/commands/developer/scan.rs [hunk 13]
```diff
@@ -0,0 +1,124 @@
...
+use snarkvm::prelude::{Block, Plaintext, Record, ViewKey};
+
+use anyhow::{bail, Result};
+use clap::Parser;
+use std::str::FromStr;
+
+// TODO (raychu86): Figure out what to do with this naive scan. This scan currently does not check if records are already spent.
+/// Scan the snarkOS node for records.
+#[derive(Debug, Parser)]
+pub struct Scan {
```
Comment (iamalwaysuncomfortable) [inline cli/src/commands/developer/scan.rs:28]: Most unspent records will likely be more towards the newest blocks than the oldest ones, perhaps we can put in options for scanning for a max balance (for example, just enough to make a transfer) and scanning in reverse.

Here are some thoughts in code about that: https://github.com/AleoHQ/snarkOS/pull/2237/files

File: cli/src/commands/developer/scan.rs [hunk 14]
```diff
@@ -0,0 +1,124 @@
...
+    #[clap(short = 'v', long, help = "The view key used to scan the snarkOS node")]
+    pub view_key: String,
+
+    #[clap(long, help = "The block to start scanning", default_value = "0")]
+    pub start: u32,
+    #[clap(long, help = "The block to stop scanning")]
+    pub end: Option<u32>,
+
+    #[clap(long, help = "The URL to scan blocks from")]
+    endpoint: String,
```
Comment (iamalwaysuncomfortable) [inline cli/src/commands/developer/scan.rs]:
```suggestion
    /// The view key used to scan the snarkOS node.
    #[clap(short, long)]
    pub view_key: String,

    /// The block height to start scanning at
    #[clap(long, default_value = "0")]
    pub start: u32,
    
    /// The block height to stop scanning
    #[clap(long)]
    pub end: Option<u32>,

    /// The endpoint to scan blocks from
    #[clap(long)]
    endpoint: String,
```

File: node/rest/src/routes.rs [hunk 15]
```diff
@@ -153,21 +153,21 @@ impl<N: Network, C: ConsensusStorage<N>, R: Routing<N>> Rest<N, C, R> {
             .and(with(self.ledger.clone()))
             .and_then(Self::find_block_hash);
 
-        // GET /testnet3/find/deploymentID/{programID}
-        let find_deployment_id = warp::get()
-            .and(warp::path!("testnet3" / "find" / "deploymentID" / ..))
+        // GET /testnet3/find/transactionID/deployment/{programID}
+        let find_transaction_id_from_program_id = warp::get()
+            .and(warp::path!("testnet3" / "find" / "transactionID" / "deployment" / ..))
```
Comment (iamalwaysuncomfortable) [inline node/rest/src/routes.rs]: I %100 support this rename

**Review by iamalwaysuncomfortable — COMMENTED** [review]
Comments on the following things:
* Clap usage
* Scan command features
* Ideas of how this might be able to integrate with the Aleo CLI.

This PR is absolutely amazing work that provides an authoritative way to do deploy/execute. Thanks so much for creating it!

File: cli/src/commands/developer/mod.rs [hunk 16]
```diff
@@ -136,14 +139,28 @@ impl Developer {
...
-                        bail!("❌ Failed to deploy '{}' to {} ({})", operation.bold(), &endpoint, error)
-                    }
-                    Transaction::Execute(..) => {
-                        bail!("❌ Failed to broadcast execution '{}' to {} ({})", operation.bold(), &endpoint, error)
+                Err(error) => {
+                    let error_message = match error {
+                        ureq::Error::Status(code, response) => {
+                            format!("(status code {code}: {:?})", response.into_string()?)
+                        }
+                        ureq::Error::Transport(error) => format!("({})", error.to_string()),
```
Comment (ljedrz) [inline cli/src/commands/developer/mod.rs]:
```suggestion
                        ureq::Error::Transport(error) => format!("({})", error),
```

File: cli/src/commands/developer/mod.rs [hunk 17]
```diff
@@ -136,14 +139,28 @@ impl Developer {
...
+                    };
+
+                    match transaction {
+                        Transaction::Deploy(..) => {
+                            bail!("❌ Failed to deploy '{}' to {} {}", operation.bold(), &endpoint, error_message)
```
Comment (ljedrz) [inline cli/src/commands/developer/mod.rs]:
```suggestion
                            bail!("❌ Failed to deploy '{}' to {}: {}", operation.bold(), &endpoint, error_message)
```

File: cli/src/commands/developer/mod.rs [hunk 18]
```diff
@@ -136,14 +139,28 @@ impl Developer {
...
+                        }
+                        Transaction::Execute(..) => {
+                            bail!(
+                                "❌ Failed to broadcast execution '{}' to {} {}",
```
Comment (ljedrz) [inline cli/src/commands/developer/mod.rs]:
```suggestion
                                "❌ Failed to broadcast execution '{}' to {}: {}",
```

**Review by howardwu — APPROVED** [review]
LGTM

**Pull request merged**